### ISSUE TITLE
Not the same conditional input

### ISSUE THREAD

**Issue by clagarbar3** [glederrey/DATGAN#4]
Hi!  @glederrey 
I'm working on a project for my master's thesis, the idea is generating a synthetic tabular relational database. I know it isn't the idea of this model, but I think using conditional inputs to maintain congruence between primary and foreign keys may be a good idea. 

However, I train the model on the first table (with more categorical than continuous variables, so it takes the WGAN loss) and indicate that the column associated with the primary key (with unique values) is a conditional entry ( DATGAN(conditional_inputs=['id']) ) . When I sample by giving as inputs the column associated to the primary key in the real table ( gan.sample(len(df), inputs=df[gan.conditional_inputs]) ), the synthetic table yields a primary key column that does not match column given as input (even repeats id's).

I don't know why, i understood that the purpose of conditional inputs (ciDATGAN) was to complete a given table, so that the feeder table matches synthetic and real data. I thought it might be because the model is not training well (?) I attach a plot with the losses over 200 epochs ("loss_real", "loss_fake" and "loss" from the training .json for the discriminator and "loss" for generator). I am surprised that the discriminator loss starts and stays practically at 0.

I would be very grateful for the help, I am presenting the project soon and I am very puzzled. Greetings, 
Claudia.

P.S: I had to put the column 'id' as a source node in the graph to at least connect to another variable, so it only points to one. Also, my dataset is limited and only has 2913 rows.

![Training Plot](https://github.com/user-attachments/assets/4dc4f886-2343-4914-b357-523c6146f42d)

**Comment by glederrey** [comment]
Hi,

I cannot check in more details right now. (I might be able to find some time tomorrow) And I'm not really maintaining this code anymore since I finished my PhD some time ago. =) But if I can help you, I will try to.

My first comments are the following: 
- your dataset is very small. It might be too small for a deep learning model. Maybe, it would be better to use a statistical model instead.
- you're training the model for too long. No need to go further than 50 epochs since it does not improve. 
- if you want to use conditional inputs, if I remember correctly, you need to have these columns in the training set as well. Is that the case? 

Is it possible for you to share the code you're using to train the model and to generate the data? It might help me find what's the problem

**Comment by clagarbar3** [comment]
You don't know how happy this response makes me, thank you for taking the time!! I take this opportunity, since I didn't do it before, to congratulate you for your PhD thesis, honestly it's super innovative for it's time. I answer all your questions:

1.  I know the data is small for DL, the problem is that I am modeling patient records, so the parent table (patients) only has 2913 unique rows while the rest of the tables (such as patient medical visits) are up to 400,000. As much as it pains me as a statistician, statistical modeling is not an option since the thesis is on logic, computation and AI, specifically generative modeling. 
2.  Correct, the id's are in the training set as a categorical variable (with as many classes as rows in the table). I know it doesn't make much sense to have a variable with so many categories but it is a way to make sure that only valid identifiers are generated.

 I also tried changing the hyperparameters (learning_rate, num_gen_rnn, num_gen_hidden, num_dis_hidden and noise) to make the generator stronger so that the discriminator would not dominate, but the loss graph during training did not change much. With WGGP it went further down to 0, with SGAN (without conditional inputs) it did change and did not converge to 0, it stayed around 0.6-0.8.


Of course I can attach the code, but not the whole data set for privacy reasons:

import tensorflow as tf # version: 2.8.0
import datgan # 2.1.7
from datgan import DATGAN
import pandas as pd # 2.0.3
import numpy as np# 1.26.4
import networkx as nx # 3.3

A little bit of information about the dataset:
person_id                       int64   # numbers but unique id for each person
gender_concept_id               int64    # 2 categories
year_of_birth                   int64  # 78 categories
month_of_birth                  int64  # 12 categories
day_of_birth                    int64  # 31 categories
year_of_death                   int32 # 9  categories (the year of death or 0 if the person still alive)
month_of_death                  int32  # 13  categories (the month of death or 0 if the person still alive)
day_of_death                    int32  # 32  categories (the day of death or 0 if the person still alive)
n_visit_occurrence              int64  # number of medical visits (rows in other table)
n_drug_exposure                 int32 # periods of drug exposure communicated or not during a medical visit (rows in other table)
n_drug_exposure_during_visit    int32 # periods of drug exposure communicated during a medical visit  (rows in other table)
n_procedure_occurrence          int32 # number of medical procedures  (rows in other table)
n_condition_occurrence          int64 # number of medical diagnostics (rows in other table)
dtype: object

metadata_person = {
     'n_visit_occurrence': {
        'type': 'continuous', 
        'bounds': [1, 1600],
        'discrete': True
    },
     'n_drug_exposure': {
        'type': 'continuous', 
        'bounds': [0, 265], 
        'discrete': True
    },
     'n_drug_exposure_during_visit': {
        'type': 'continuous', 
        'bounds': [0, 215],
        'discrete': True
    },
    'n_procedure_occurrence': {
        'type': 'continuous', 
        'bounds': [0, 115], 
        'discrete': True
    },
    'n_condition_occurrence': {
        'type': 'continuous', 
        'bounds': [1, 425], 
        'discrete': True
    }
    }

for c in person.columns:
    if c not in metadata_person.keys():
        metadata_person[c] = {'type': 'categorical'}

graph = nx.DiGraph()
graph.add_node("gender_concept_id")  # single node
graph.add_edges_from([
    ("person_id", "n_visit_occurrence"), # need to add this link for conditional input
    ("year_of_birth", "month_of_birth"),
    ("month_of_birth", "day_of_birth"),
    ("year_of_birth", "day_of_birth"),
    ("year_of_birth", "year_of_death"),
    ("year_of_birth", "month_of_death"),
    ("year_of_birth", "day_of_death"),
    ("year_of_birth", "n_condition_occurrence"),
    ("year_of_birth", "n_procedure_occurrence"),
    ("year_of_birth", "n_visit_occurrence"),
    ("year_of_birth", "n_drug_exposure"),
    ("year_of_death", "month_of_death"),
    ("year_of_death", "day_of_death"),
    ("month_of_death", "day_of_death"),
    ("year_of_death", "n_condition_occurrence"),
    ("month_of_death", "n_condition_occurrence"),
    ("day_of_death", "n_condition_occurrence"),
    ("year_of_death", "n_visit_occurrence"),
    ("month_of_death", "n_visit_occurrence"),
    ("day_of_death", "n_visit_occurrence"),
    ("year_of_death", "n_procedure_occurrence"),
    ("n_drug_exposure", "n_drug_exposure_during_visit"),
    ("n_condition_occurrence", "n_drug_exposure"),
    ("n_condition_occurrence", "n_drug_exposure_during_visit"),
    ("n_condition_occurrence", "n_procedure_occurrence"),
    ("n_condition_occurrence", "n_visit_occurrence"),
    ("n_procedure_occurrence", "n_visit_occurrence"),
    ("n_procedure_occurrence", "n_drug_exposure"),
    ("n_drug_exposure", "n_visit_occurrence"),
    ("n_drug_exposure_during_visit", "n_visit_occurrence")
])

try:
    cycle = nx.find_cycle(graph, orientation="original")  # For Directed Acylic Graph
    print("Graph has a cycle:", cycle)
except nx.exception.NetworkXNoCycle:
    print("Graph has no cycles")

datgan.advise(person, graph, plot_graphs=True)


output_folder = './salida person/'
gan_person = DATGAN(output=output_folder, num_epochs=200,  batch_size=224, conditional_inputs=['person_id'])
gan_person.preprocess(person, metadata_person, preprocessed_data_path='./person preprocesada')
history = gan_person.fit(person, metadata_person, graph, preprocessed_data_path='./person preprocesada')

#Output during train:
#Preprocessed data have been loaded!
#Start training DATGAN with the WGAN loss (17/10/2024 16:46:13).
#Training DATGAN: 100%|██████████| 200/200 [16:24<00:00,  4.92s/it]
#DATGAN has finished training (17/10/2024 17:02:38) - Training time: 16 minutes and 24 seconds

samples_person = gan_person.sample(len(person), inputs=person[gan_person.conditional_inputs])

len(set(samples_person.person_id)) #only 1893 unique id's

**Comment by glederrey** [comment]
Hi, 

Thanks for the kind words. =) Even if this project worked well in the past, I'm pretty sure that Transformers would make the model much better. But I don't have the courage to work on this. =)

I totally understand that you cannot share your data. You already shared enough information that should allow me to help you a bit. 

I plotted the graph that you have here:

![newplot](https://github.com/user-attachments/assets/930a6241-b7fb-4999-bb70-eebe85de3713)

The first thing I see is that it's very concentrated on the variable `year_of_birth`. If I remember correctly how the model performed, it tend to underperform. So, you should try to remove some links. For example, are you sure the `n_drug_exposure` is really caused by the `year_of_birth`? Maybe there's a link. But maybe there's a variable before that should chain these. 

Next advice, still about the graph, I would prune the dataset/graph to remove all the causal links that should not be there. For example, I'm not sure if the following variables are needed:
  - `month_of_birth` & `day_of_birth`: These two are linked to the `year_of_birth`. I don't think they should since the year you're born should not affect the day, nor the month. In addition, I don't think they would impact your data. If you need to generate these as well, I would use a standard random generation. 
  - `month_of_death` & `day_of_death`: Similarly, you could generate these two later on.

I would strongly advise you to manually draw the graph. You should really start with one variable. Then, you can, for example, check which variables are the most correlated to your first variables. If there's no correlation, there's no causality. Thus, there shouldn't be any link. It might be a good idea to start with `year_of_birth`. But you should also try to draw some graphs starting from other variables.

Now, let's talk about the variable `person_id`. In my opinion, the ID shouldn't be included in the graph, nor in the generation process. If I understand correctly, you want to augment your dataset, right? So, if you want to augment it, you want to have new "synthetic" patients. Thus, these should have new IDs. This could never happen if the `person_id` is given in the model as a categorical variable because the model will only generate existing categorical variables. It cannot generate new categories. (We would need LLMs for that =P)

 On the other hand, if you want to generate new "scenarios" for your existing patients, I would do the following:
- do not include the `person_id` (it does not cause any of the other variables), but set it aside in a table that links the `person_id` with the date of birth and date of death.
- Train the model with the pruned graph
- For the conditional part, set the date of birth and the date of death as the conditional variables and let the model generate the other variables. 
- You can, then, based on the two dates, give back the `person_id` for the multiple scenarios. 

Also, some small other advice: I would add a boolean variable for the `year_of_death` named `alive`. That way, you can make sure that the model understands that the `year_of_death` should be 0 (and you can enforce it after generation if needed) when `alive` is False. 

A few other things you could try to improve the model:
- Make the NN smaller (set the different sizes of the network to smaller values, at least half, just to try)
- Train the model with smaller batch size 

But in the end, I definitely think that the main problem comes from your data and the definition of the graph. As explained in the paper, each link is directed to simulate causation. There should only be a link between A and B if A influences B (so, at least high pos/neg correlation). So, start with upgrading the graph as I explained. You can definitely compute the correlations between the variables to help you. 

I hope my message will help you. Sorry for answering this late.

**Comment by clagarbar3** [comment]
Hi again! @glederrey 

Thanks a lot for the answer, of course it's helpful :) Maybe the Self-Attention mechanism could provide better results, but I don't know if like you, I'm a bit saturated of seeing Transformers in eeevery project.

On the other hand, I had already tried many of the things you mentioned:

-  I included a boolean variable `death` that indicated if the person was alive (0) or dead (1), the problem is that the model wasn't able to capture the relationship of that variable with the year, month and day of death so it rarely indicated all those columns as a 0 (i.e. the person is dead). I will try including that variable again and exclusively the year of death as you comment.  I have also tried to unify year, month and day of death in a continuous variable `days_to_death` (0 if the person is still alive) but the model doesn't return people with value 0 in that column.

- You are absolutely right about the `person_id`, if I add it as a categorical variable I wouldn't be able to generate new patients, I must use several columns to condition the input. However I am still a bit puzzled about the “conditional input”. In the paper you mention that those inputs are copied directly to the synthetic set (i.e. I enter the columns `year_of_birth` and `year_of_death` and the model outputs the rest of the columns), but when I provide the inputs the sampled set does not have the exact columns I gave it, so might appear individuals that I can't find in the mapping table. In fact, I have seen in this repository example the `DATGAN_ci.csv` file does not have the same `age`, `gender`, and `trip_purpose` columns as the original `CMAP.csv` set. Perhaps I misunderstood the concept of conditional input?

- Finally, the idea is not so much to augment the data, but to be able to generate data from multiple tables that preserve the primary and foreign keys between them, which is why conditional input is so relevant. Moreover, although it seems that the network is centered on `year_of_birth` the idea is really to estimate the number of visits `n_visit_occurrence` (that's why is a sink node), cardinality that I need later to generate the table of visits (as many as visits each patient has).

To clarify, you mention that the model tends to perform poorly, so do you recommend including as few variables as possible? When I read the paper I thought the opposite, as well as adding as many connections in the graph as possible.

**Comment by glederrey** [comment]
Hi, 

I will try to answer all your questions. 

- About the data itself, if you have very few "alive" patients, the model will definitely fail to represent them. What you could do is to have a boolean variable to `death` and if the model generate this as `False`, you can set the other values to 0 or something like this. Honestly, I never tried to train the model with variables like this. That's possible the model does not understand how these type of variable work.
-  The `sample` function of the `DATGAN` model has multiple parameters. If you want to find exactly the same values as the input, you should add `randomize=False`. The model will, then, sample without replacement. I just saw that it's not in the README file. Sorry about that. I will update it now. Something like this:
```
samp = datgan.sample(len(tmp), inputs=tmp, randomize=False)
```
- Ahhhh, so you want to generate other data for the patients. Meaning that you absolutely want to keep the `person_id`, right? If it's the case, then you add it as a categorical variable. However, I'm not so sure how it will work in the network since it's and ID and it should not be correlated with the other variables. If it doesn't work, create a new "key" composed of the `person_id`, the year of birth and year of death. Then, you don't include the `person_id` in the graph, but whenever you generate someone with given year of birth and year of death, you can manually assign the `person_id`

I won't have access to my computer until Sunday evening. But if you have more questions, I'll answer you then. 

Have a nice weekend

**Comment by clagarbar3** [comment]
Hii!! @glederrey 
 
thank you very much for the help. When you mentioned the `randomize=False` method I was surprised because `DATGAN.sample` didn't have that argument (I'm using version 2.1.7). I searched the rest of the versions available in Python 
[here](https://pypi.org/project/datgan/#history) and they don't mention anything about that argument. I had to copy the function from this repository, I comment it in case someone else might find it useful. On the other hand, I would like to ask you two things:

- I'm now generating an even bigger table (more than 200,000 rows, reduced to a training set of 70,000). In general with the `WGAN` and `WGGP` functions the losses associated to the discriminator are kept from the beginning at 0, which could be a failure in the convergence (?). No matter how much I modify the hyperparameters I can't solve it (with the smaller table the same thing was also happening). In general, I can't get the network to capture the Normalized Mutual Information Matrix (NMI) of the real data.

- I have noticed, that if I train for few epochs (3-5) generating the sample of 70.000 rows (with conditional inputs) is relatively fast, however the more epochs I train (10-15) getting the sample takes twice as long, does that make any sense?

**Comment by glederrey** [comment]
Hi, 

Sorry, as I said, I haven't maintained this correctly since I finished my PhD. I must have fucked this up. 

- Honestly, I'm really not sure. I guess, it depends on the data. Is it really 0.0000 or is it an issue of printing the values?
- Mmmmhhhh, it shouldn't imo. At least, not twice as long. 

Sorry for not helping much on this. Without trying myself (and a clone of myself to have more time available ^^), I don't really know how to help you. =(

**Comment by clagarbar3** [comment]
Don't worry, it's totally understandable that you don't have the repository updated, you are doing enough for me :)

Regarding the losses it's strange, with the smaller table I only managed to get training and improvement with `SGAN` for 100-150 epochs (simple and short time because it was little data), the NMI matrix and data correlations replicated very well. The plot was similar to this:

![SGAN small data](https://github.com/user-attachments/assets/90282f95-256b-4504-af14-f227ec8ea121)

Now that I have a lot more data, I still can't get the model to train with `WGAN` or `WGGP`, it replicates only slight correlations. The plot for `WGAN` is something like this, discriminator with losses like 0.01 or in some cases a constant straight line at a value practically 0 for synthetic data and a bit higher for real:

![WGAN larger data](https://github.com/user-attachments/assets/969f3c1e-fd42-4ddb-a735-4c90a8ad3109)

I was thinking that maybe I should train for more epochs (?) it puzzles me that the loss of the discriminator doesn't fluctuate and from the beginning is soooo low. I thought the discriminator might be dominating and increased the noise in the label smoothing and the number of units in the generator layers, but nothing. I also tried increasing the input noise vector of the generator.

**Comment by glederrey** [comment]
Hi, 

There's no need to train for more epochs. The models are not improving. 

Could you try to train with a bigger batch size? I wonder if this could help.
